File: scrapper/offer_scrapper.py
import os
import asyncio
import random
from playwright.async_api import async_playwright
from supabase import create_client, Client
from dotenv import load_dotenv

load_dotenv(dotenv_path='supabase.env')
load_dotenv(dotenv_path='credentials.env', override=True)

SUPABASE_URL = os.getenv('SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_ANON_KEY')
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

IG_USER = os.getenv('IG_USER')
IG_PASS = os.getenv('IG_PASS')

async def human_wait(min_s=1.0, max_s=3.0):
    delay = random.uniform(min_s, max_s)
    await asyncio.sleep(delay)

async def scrape_instagram():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False)
        context = await browser.new_context()
        page = await context.new_page()

        # Login
        try:
            await page.goto("https://www.instagram.com/accounts/login/")
            await human_wait(2, 4)

            await page.fill("input[name='username']", IG_USER)
            await human_wait(1, 3)

            await page.fill("input[name='password']", IG_PASS)
            await human_wait(1, 4)

            await page.click("button[type='submit']")
            await page.wait_for_url("https://www.instagram.com/accounts/onetap/?next=%2F", timeout=60000)

            await page.click("text='Not now'", timeout=60000)
            await human_wait(2, 4)
        finally:
            await page.goto("https://www.instagram.com/explore/tags/oferta/")
            await page.wait_for_selector("a[href^='/p/']", timeout=60000)
            links = await page.eval_on_selector_all("a[href^='/p/']",
                                                    "elements => elements.map(e => e.href)")
        print(f"Se encontraron {len(links)} publicaciones.")

        # Scroll
        for _ in range(6):
            await page.mouse.wheel(0, random.randint(500, 1500))
            await human_wait(2, 5)

        # Extraer imagen y descripción de cada publicación
        for url in links:
            await page.goto(url)
            await human_wait(1, 2)
            try:
                img_url = await page.get_attribute("img[src]", "src")
                description = await page.eval_on_selector("span", "el => el.innerText")
            except Exception as e:
                print(f"Error extrayendo datos de {url}: {e}")
                img_url = None
                description = None

            offer_data = {
                "url": url,
                "imagePath": img_url,
                "description": description,
            }
            supabase.table("offer").insert(offer_data).execute()
            await human_wait(0.5, 1.5)

        # Insertar en Supabase
        for url in links:
            offer_data = {"url": url, "imagePath": img_url, "description": description}
            supabase.table("offer").insert(offer_data).execute()
            await human_wait(0.5, 1.5)

        await browser.close()

def save_to_supabase(offer_data):
    if not offer_data:
        print("No hay datos nuevos para guardar.")
        return
    print("Guardando datos en Supabase...")
    response = supabase.table("offer").insert(offer_data).execute()
    if response.get("error"):
        print("Error al guardar en Supabase:", response["error"]["message"])
    else:
        print("¡Datos guardados exitosamente en Supabase!")

def main():
    try:
        scraped_data = asyncio.run(scrape_instagram())
        save_to_supabase(scraped_data)
    except Exception as e:
        print(f"Ocurrió un error: {e}")

if __name__ == '__main__':
    main()

File: lib/scrapper/offer_scrapper.ts
import { chromium, Page } from 'playwright';
import { insertOffer } from '@/lib/supabase/repository'
import { Offer } from '@/domain/interface'
import { config } from 'dotenv';
config({ path: 'data.env' });

const IG_USER = process.env.IG_USER!;
const IG_PASS = process.env.IG_PASS!;

function humanWait(min = 1000, max = 3000) {
    const delay = Math.random() * (max - min) + min;
    return new Promise(res => setTimeout(res, delay));
}

async function loginInstagram(page: Page) {
    await page.goto('https://www.instagram.com/accounts/login/');
    await humanWait(2000, 4000);
    await page.fill("input[name='username']", IG_USER);
    await humanWait(1000, 3000);
    await page.fill("input[name='password']", IG_PASS);
    await humanWait(1000, 4000);
    await page.click("button[type='submit']");
    await page.waitForURL('https://www.instagram.com/accounts/onetap/?next=%2F', { timeout: 60000 });
    await page.click("text=Not now", { timeout: 60000 });
    await humanWait(2000, 4000);
}

async function scrapeInstagramOffers() {
    const browser = await chromium.launch({ headless: false, args: ['--no-sandbox'] });
    const context = await browser.newContext();
    const page = await context.newPage();
    await loginInstagram(page);

    await page.goto('https://www.instagram.com/explore/tags/oferta/');
    await page.waitForSelector("a[href^='/p/']", { timeout: 60000 });

    for (let i = 0; i < 6; i++) {
        await page.mouse.wheel(0, Math.floor(Math.random() * 1000) + 500);
        await humanWait(2000, 5000);
    }

    const links: string[] = await page.$$eval("a[href^='/p/']", els => els.map(e => (e as HTMLAnchorElement).href));
    console.log(`Se encontraron ${links.length} publicaciones.`);

    const offers: any[] = [];

    for (const url of links) {
        await page.goto(url);
        await humanWait(1000, 2000);
        let img_url: string | null = null;
        let description: string | null = null;
        try {
            img_url = await page.getAttribute("img[src]", "src");
            description = await page.$eval("span", el => el.textContent || '');
        } catch (e) {
            console.log(`Error extrayendo datos de ${url}: ${e}`);
        }

        offers.push({
            url,
            imagesPath: img_url ? [img_url] : [""],
            description,
            user_id: ""
        });
        await humanWait(500, 1500);
    }

    await browser.close();
    return offers;
}

async function saveOffersInSupabase(offers: any[]) {
    for (const offer of offers) {
        try {
            await insertOffer(offer);
            console.log('¡Oferta guardada exitosamente en Supabase!', offer.url);
        } catch (error) {
            console.log('Error al guardar en Supabase:', (error as Error).message);
        }
    }
}

async function main() {
    const offers = await scrapeInstagramOffers();
    await saveOffersInSupabase(offers);
}
console.log('About to run main...');
main().catch(e => console.error(`Ocurrió un error: ${e}`));